Implemented new Method to Calculate Beams with Hinges

mvg2002 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
This PR solves the problem described in issue #26570. 


#### Brief description of what is fixed or changed
Implemented a new way to calculate beams with hinges using singularity functions. Using this new method beams with multiple hinges can be calculated. 

#### Other comments
As this is a large pull request where multiple things change at once I will go though it point by point. 

1. Added method `apply_hinges`.
This method places a hinge anywhere in the beam using a singularity function of order -3. 
It introduces a new unknown: the rotation of the hinge. 
It introduces a new boundry condition: bending moment is 0 at the hinge. 

2. Added new property `bc_bending_moment` which shows the bending moment boundry conditions. 
The setter of this property also adds the boundry conditions to `_boundry_conditions`.

3. Added the new unknowns and boundry conditions to `solve_for_reaction_loads`.
In this method the new unknowns and boundry conditions are added which gives this method the possibility to calculate beams with multiple hinges. 

4. Changed the way the `join` method handles hinges. 
The `join` method used to handle hinges by setting `self._composite_type` to hinge. Now that the `apply_hinge` method is there it can apply a hinge on the new beam at the place where the beams are joined. 

5. Changed `self._composite_type` to `self._joined_beam`
As hinges are handled in a different way now it doesn't matter anymore if a beam is joined by a hinge or fixed. For some functions it is still needed to know if a beam is joined or not. That is why `self._composite_type` is changed to `self._joined_beam`. 

6. Removed `_solve_hinge_beams`.
As hinge beams can now be calculated in `solve_for_reaction_loads` this `_solve_hinge_beams` method became obsolete. 

7. Small change in `draw` method, so that it can draw multiple hinges if necessary.

8. Added an error message when trying to do I.L.D. calculations on a beam with hinges. 
This was not, and still is not, implemented. This can be done at a later time. 

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* physics.continuum_mechanics
    * Implemented hinges in a new way which makes it possible to calculate beams with multiple hinges. 

<!-- END RELEASE NOTES -->

Commit:
Implemented apply_hinge function (all tests pass)

Commit:
Remove code that has been replaced and update join method.

Commit:
Add NotImplementedError for I.L.D. calculations.

Commit:
Update docstrings

Commit:
Small improvement in handling hinge symbols.

Commit:
Doctest fix

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -491,6 +501,55 @@ def apply_support(self, loc, type="fixed"):
+
+        """
+        loc = sympify(loc)
+        rotation_jump = Symbol('P_'+str(loc))
```
Is this a symbol that the user may need to access? I don't see where it appears in the above example. In general, we should not create symbols internally, users should have to pass them in so that they have access to the variable we should make them accessible.

mvg2002 replied:
In my opinion this is not a symbol that the user may need to access. This symbol describes the unknown rotation of the hinge in the beam. This is normally not something the user is interested in. It is necessary to create the symbol to describe the unknown rotation jump that is solved when the user calls `solve_for_reaction_loads`. The symbol is assigned a value in this method and that value is substituted into the load equation. 

Of course it is possible to handle this unknown in the same way the unknown reaction forces are handled in `apply_support`. The reason I have chosen not to do this in this PR is because this would make using the method a bit more difficult. This way the user would have to undertake more steps and the chance of entering the wrong `*reactions` into `solve_for_reaction_loads` would increase. 

I have taken both sides into consideration and thought this was the best way to handle this. As in my experience, you are almost never interessed in the value of rotation in a hinge.

mvg2002 replied:
Now that I think about it there might be a way to have acces to the symbol but not have to pass it into  `solve_for_reaction_loads`. I will try to implement this and push when it works. This will probably be later today.

mvg2002 commented:
I will also take another look at the doctests that keep failing.

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -409,13 +419,13 @@ def join(self, beam, via="fixed"):
         if via == "hinge":
             new_beam = Beam(new_length, E, new_second_moment, x)
-            new_beam._composite_type = "hinge"
-            new_beam._hinge_position = self.length
```
Are these not somehow needed in the current implementation? What are the consequences of deleting them?

mvg2002 replied:
The new code is:
``` 
        if via == "hinge":
            new_beam = Beam(new_length, E, new_second_moment, x)
            new_beam._joined_beam = True
            new_beam.apply_hinge(self.length)
            return new_beam
```

This does the same as the old code, but in a different way. In the new way hinges can be placed anywhere in a beam instead of just the place where two beams are connected. The statement `new_beam._joined_beam = True` makes sure that it is known that the beam consists of two different parts that can have different E and I. This is necessary for slope and deflection calculations. The way of joining the two beams, fixed or by hinge, doesn't matter anymore because hinges are handled in a different way. 

The `new_beam.apply_hinge(self.length)` line places a hinge exacly at the place where the two beams are connected. and in the `apply_hinge` method it keeps track of all locations where hinges are placed. This replaceses the old line of `new_beam._hinge_position = self.length`. 

The main reason this is possible is because in the new way beams with and without hinges can be calculated in the same way. It doesn't matter anymore if there are hinges, so it is no more necessary to specify that the `_composite_type` is a hinge.

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -710,141 +769,6 @@ def applied_loads(self):
-
-        self._hinge_beam_slope = slope_1*SingularityFunction(x, 0, 0) - slope_1*SingularityFunction(x, l, 0) + slope_2*SingularityFunction(x, l, 0)
-        self._hinge_beam_deflection = def_1*SingularityFunction(x, 0, 0) - def_1*SingularityFunction(x, l, 0) + def_2*SingularityFunction(x, l, 0)
-
```
It looks like you are removing the prior functionality in favor of the new way. If this changes any thing that is public and user facing, then we have to follow the deprecation policy. Does this change anything for the user?

Also, does swapping for the new way solve all prior hinge problems that worked, plus new ones? i.e. does it disable anything that worked before?

mvg2002 replied:
This new way doesn't change anything for the user. The user still uses `solve_for_reaction_loads(*reactions)` to calculate reaction forces and does not need to change anything about their input. Not for a beam without hinges nor for a beam with hinges. This can also be seen in the test file. All tests of beams with and without hinges still pass without any changes to the tests. 

For the second question. This new way solves the problem stated in issue #26570 and it keeps all previous functionalities that were there before. But most importantly it adds a new functionality: beams can have multiple hinges. In addition to this hinges can be added easier to a beam than before by using the `apply_hinge` method, without removing the old way of adding hinges by using the `join` method.

moorepants commented on sympy/physics/continuum_mechanics/tests/test_beam.py:
```diff
@@ -509,6 +510,38 @@ def test_apply_support():
+        + 31 * SingularityFunction(x, 10, 3) / 7200 - SingularityFunction(x, 10, 4) / 4800
+        - 5137 * SingularityFunction(x, 12, 1) / 5184 + SingularityFunction(x, 15, 3) / 800
+        + SingularityFunction(x, 15, 4) / 4800)
+    assert b.deflection().expand() == expected_deflection.expand()
```
It is important to test a fully symbolic system to make sure the symbolics function as well as a numeric example.  Different bugs may arise from both tests.

mvg2002 commented:
I understand that this is a big change. Especially the removal of the `_solve_hinge_beams` method. I will look out for problems with current functionalities and make sure there are no changes for current users. 

For now I will continue with the following 3 points:
1. Change the `apply_hinge` method in such a way that the rotation jump is also accesible to the user. 
2. Add a symbolic test for `apply_hinge`.
3. Fix the problem in the doctest.

moorepants commented:
> Especially the removal of the _solve_hinge_beams method.

That's fine. It is a private method so it can be removed without deprecation if user facing functionality doesn't change.

mvg2002 commented:
After looking at excersice 8 of the [beam_problems.rst](https://github.com/sympy/sympy/blob/master/doc/src/modules/physics/continuum_mechanics/beam_problems.rst) file, I see that there is a change in output between the new function and the old one. This change happens because the hinge is added to the load equation as a singularity function of order -3. This gives much easier and shorter slope and deflection equations, as I will show below here. But the problem of this change in equations is that the output of for example `b.load` and `b.slope` are different than before. That is why the doctest kept failing. I will show the difference below here.

The original output of `b.load`:
```
             -2          -1               -1                                -1
      P⋅l⋅<x>     3⋅P⋅<x>     5⋅P⋅<-l + x>                 -1   P⋅<-4⋅l + x>
    - ───────── + ───────── - ────────────── + P⋅<-3⋅l + x>   - ──────────────
          4           4             4                                 2
```

The new output of `b.load`:
```
            2           -3          -2          -1               -1                                -1
      13⋅P⋅l ⋅<-2⋅l + x>     P⋅l⋅<x>     3⋅P⋅<x>     5⋅P⋅<-l + x>                 -1   P⋅<-4⋅l + x>
    - ──────────────────── - ───────── + ───────── - ────────────── + P⋅<-3⋅l + x>   - ──────────────
               48                4           4             4                                 2
```

It is visible that the load equation became a bit longer because of the addition of the singularity function describing the hinge. The main adventage of this becomes visible when looking at the slope and deflection equation. The slope equation is shown below. 

Original output of `b.slope`:
```
    ⎛     2               2               2               2⎞               ⎛         1          2               2               2⎞        ⎛         1          2               2               2⎞
    ⎜5⋅P⋅l    P⋅<-2⋅l + x>    P⋅<-3⋅l + x>    P⋅<-4⋅l + x> ⎟           0   ⎜  P⋅l⋅<x>    3⋅P⋅<x>    5⋅P⋅<-l + x>    P⋅<-2⋅l + x> ⎟    0   ⎜  P⋅l⋅<x>    3⋅P⋅<x>    5⋅P⋅<-l + x>    P⋅<-2⋅l + x> ⎟           0
    ⎜────── - ───────────── + ───────────── - ─────────────⎟⋅<-2⋅l + x>    ⎜- ──────── + ──────── - ───────────── + ─────────────⎟⋅<x>    ⎜- ──────── + ──────── - ───────────── + ─────────────⎟⋅<-2⋅l + x>
    ⎝  48           4               2               4      ⎠               ⎝     4          8             8               4      ⎠        ⎝     4          8             8               4      ⎠
    ──────────────────────────────────────────────────────────────────── + ──────────────────────────────────────────────────────────── - ───────────────────────────────────────────────────────────────────
                                    E⋅I                                                                E⋅I                                                                E⋅I
```

The new output of `b.slope`: 
```
     ⎛      2           0          1          2               2               2               2⎞
     ⎜13⋅P⋅l ⋅<-2⋅l + x>    P⋅l⋅<x>    3⋅P⋅<x>    5⋅P⋅<-l + x>    P⋅<-3⋅l + x>    P⋅<-4⋅l + x> ⎟    
    -⎜─────────────────── + ───────── - ──────── + ───────────── - ───────────── - ─────────────⎟
     ⎝       48               4          8             8               2               4      ⎠
───────────────────────────────────────────────────────────────────────────────────────────────────
                                                 E⋅I
```

This is a change that user can experience. If they calculate anything with these functions the result will be the same, but the function as a whole will look different. If a user plots the slope or uses `b.slope.subs(x, 3l)` the result will be the same, but as shown above the output of `b.slope` will be different. 

My question is if this is something that has to follow depreciation policy. I saw that something has to be depreciated if it's backwards incompatible and user code would stop working after the change. I'm not sure if that is the case here. Knowing all that is stated above here, is it necessary for this code do follow depreciation pollicy?

moorepants commented:
> My question is if this is something that has to follow depreciation policy.

If the output changes to a mathematically equivalent form, then no deprecation is needed. If the output changes from an incorrect answer to a correct answer, no deprecation is needed. So I think you are fine here.

mvg2002 commented:
@Tom-van-Woudenberg

Commit:
Added return statement to apply_hinge, added symbolic test, fix doctest.

Commit:
Fix code quality

Commit:
Added completely symbolic test.

Commit:
fix python-flint

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -1620,7 +1620,8 @@ def _solve_for_ild_equations(self):
         """
-
+        if self._applied_hinges:
+            raise NotImplementedError("I.L.D. calculations are not implemented for beams with hinges.")
```
As this might be a next step, could you add this as an issue?

mvg2002 replied:
I added two issues that have to do with this problem. 
Missing tests: #26602
Unable to handle hinge beams: #26601

Tom-van-Woudenberg commented:
@mvg2002, What happens if you put two hinges at the same place? In the previous implementation that was not possible (or maybe something similar with a beam of length 0).
And what happens if you add too many hinges so that a mechanism emerges? (this was already an option in the old implementation so if this is an issue you might not want to solve it)

mvg2002 commented:
> What happens if you put two hinges at the same place or at the beginning or end of a beam? In the previous implementation that was not possible (or maybe something similar with a beam of length 0).

Putting two hinges at the same place will give a `ValueError: duplicate symbols given`. This can be handled in a better way. I will add this. 

Putting a hinge at the beginning or end of a Beam gives problems, especially when there is a fixed support at the place of the hinge. I will add some error messages for this. 

Overall it will probably still be possible to break the code if a user really wants to. It will be quite difficult to catch all possible 'wrong' structures. But you are right that it is better to add the error messages right now if certain problems are known. 

> And what happens if you add too many hinges so that a mechanism emerges? (this was already an option in the old implementation so if this is an issue you might not want to solve it)

It is still possible to add too many hinges and create a mechanism. If this happens `solve_for_reaction_loads` will output `IndexError: tuple index out of range`. This is the output the method gives no matter the problem.  This is also described in the following issue comment: https://github.com/sympy/sympy/issues/21660#issuecomment-974766872. This issue was quite unrelated to what we are talking about here, so I made a new issue about this #26604.

Commit:
Add error messages + tests

Tom-van-Woudenberg commented:
> Putting two hinges at the same place will give a ValueError: duplicate symbols given. This can be handled in a better way. I will add this.
> 
> Putting a hinge at the beginning or end of a Beam gives problems, especially when there is a fixed support at the place of the hinge. I will add some error messages for this.
> 
> Overall it will probably still be possible to break the code if a user really wants to. It will be quite difficult to catch all possible 'wrong' structures. But you are right that it is better to add the error messages right now if certain problems are known.

I think these two possibilities to break the code are relevant to treat with your new implementation as the old method didn't allow you to place multiple hinges at arbitrary points. So I argee with your implementation in [43f3faf](https://github.com/sympy/sympy/pull/26594/commits/43f3fafdecb95d97d8b258643fe7d276f2f2ae07)

Commit:
Add sliding hinges

Commit:
Add error when placing moment load and hinge at same location. Add info that rotation jump is multiplied by EI.

mvg2002 commented:
In this last push is added an error message when a moment load and hinge are placed at the same location. This should not be possible. This is also tested in the test file.

I also changed the symbol for the rotation jump from `P_<loc>` to `EI_P_<loc>`. This was done because the symbol of the rotation jump is the actual rotation jump multiplied by EI. This has to do with the symbol being in the load eqaution and the actual jump in the rotation equation. While integrating the symbol gets devided by EI. This change is also added in the docstring of `self.rotation_jump`. 

These changes are already discussed with @Tom-van-Woudenberg. 
@moorepants, is the code ready for approval like this, or are there any additional changes you would like to see?

mvg2002 committed:
Merge branch 'Dev_sliding_hinge' into Dev_hinge_beams

moorepants commented:
If it is ready for review, I can look over it again. I'm busy with a report deadline at the moment so I can look first of next week.

Commit:
Fix authors

Commit:
Merge remote-tracking branch 'origin/Dev_hinge_beams' into Dev_hinge_beams

Commit:
Branch with sliding hinges

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -210,6 +215,16 @@ def reaction_loads(self):
+    @property
+    def rotation_jumps(self):
+        """ Returns the rotation jumps in hinges multiplied by the elastic modulus and second moment in a dictionary."""
+        return self._rotation_jumps
```
@mvg2002 Could you make clear that the rotation jump include EI in the variable name itself?

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -210,6 +215,16 @@ def reaction_loads(self):
+    def deflection_jumps(self): #AANPASSEN
+        """ Returns the deflection jumps in sliding hinges in a dictionary."""
+        return self._deflection_jumps
+
```
Same here, make clear that it includes EI in the symbol name

Tom-van-Woudenberg reviewed (commented):
Maybe I don't understand your code properly because you say it's included in the name somewhere else.

mvg2002 commented:
I am experiencing quite some trouble using github right now. Something went wrong while merging two personal branches, this is why some tests are failing. I will try to resolve this and then implement the feedback.

Commit:
Add EI to variable name

mvg2002 commented:
> Maybe I don't understand your code properly because you say it's included in the name somewhere else.

I already added EI to the Symbol name, so when printing `b.rotation_jumps` instead of {P_7: 120} it would give {EI_P_7: 120}. 
I agree that for clarity, it is better to also add the EI to the variable name itself. I just pushed this change.

mvg2002 commented:
As I said, I had some trouble with GitHub. Something went wrong while merging some personal branches. I accidently added changes made in a different branch to the `Dev_hinge_beams` branch. My plan was to make a pull request for these changes after finalising this pull request, but as they are here now I will explain what was added. 

In addition to all changes already explained in this pull request, sliding hinges are added. 

Sliding hinges are hinges that allow vertical movement. The boundry condition that comes with a sliding hinge is that the shear force should be zero at the location of the hinge. A real world application of such a hinge is very difficult. The reason they are used in calculations is because they can be really usefull for determining shear force in influence line diagrams and for calculations using the principle of virtual work. 

The sliding hinges can be added by using the keyword argument `type = "sliding"` in the `apply_hinge` method.

mvg2002 commented:
I will make three more changes to this pull request:
- I will add error messages when placing a point load and a sliding hinge are placed at the same location. This should not be possible. 
- I will add a warning when a sliding hinge and a rotation hinge are placed at the same location. This can be possible but is highly unusual. The warning will make sure the user takes a second look if their structure is correct.
- I will add a feature in the `draw` method to make sure sliding hinges are drawn.

Commit:
Add errors when placing hinges at wrong locations. (with tests)

Commit:
Add warning when placing rotation hinge and sliding hinge at the same location.

Commit:
Added drawing of sliding hinges.

mvg2002 commented:
The sphinx test keeps failing. The total error message is quite long, the end of the error message looks like this: 
```
LaTeX Warning: Reference `modules/physics/control/control_plots:sympy.physics.c
ontrol.control_plots.step_response_plot' on page 2238 undefined on input line 1
86035.


LaTeX Warning: Hyper reference `modules/physics/control/control_plots:sympy.phy
sics.control.control_plots.impulse_response_plot' on page 2238 undefined on inp
ut line 186035.


LaTeX Warning: Reference `modules/physics/control/control_plots:sympy.physics.c
ontrol.control_plots.impulse_response_plot' on page 2238 undefined on input lin
e 186035.


LaTeX Warning: Citation `modules/physics/control/control_plots:r733' on page 22
38 undefined on input line 186043.


LaTeX Warning: Hyper reference `modules/plotting:sympy.plotting.series.LineOver
1DRangeSeries' on page 2238 undefined on input line 186060.


LaTeX Warning: Reference `modules/plotting:sympy.plotting.series.LineOver1DRang
eSeries' on page 2238 undefined on input line 186060.

[2238]

LaTeX Warning: Hyper reference `modules/plotting:sympy.plotting.series.LineOver
1DRangeSeries' on page 2239 undefined on input line 186111.


LaTeX Warning: Reference `modules/plotting:sympy.plotting.series.LineOver1DRang
eSeries' on page 2239 undefined on input line 186111.


LaTeX Warning: Hyper reference `modules/physics/control/control_plots:sympy.phy
sics.control.control_plots.ramp_response_plot' on page 2239 undefined on input 
line 186169.


LaTeX Warning: Reference `modules/physics/control/control_plots:sympy.physics.c
ontrol.control_plots.ramp_response_plot' on page 2239 undefined on input line 1
86169.

[2239]
File: allowed-sign-conventions.png Graphic file (type bmp)
<allowed-sign-conventions.png>
[2240]
Overfull \vbox (3.06688pt too high) detected at line 186266
 []

[2241]
Overfull \vbox (3.28088pt too high) detected at line 186326
 []

[2242]

! LaTeX Error: Too deeply nested.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.186420 ...m{If type = “rotation”}\begin{itemize}
                                                   
Here is how much of TeX's memory you used:
 19139 strings out of 481804
 421068 string characters out of 5937876
 86[2744](https://github.com/sympy/sympy/actions/runs/9210431469/job/25338450894?pr=26594#step:6:2745) words of memory out of 5000000
 35641 multiletter control sequences out of 15000+600000
 548726 words of font info for 147 fonts, out of 8000000 for 9000
 15 hyphenation exceptions out of 8191
 43i,31n,67p,6436b,793s stack positions out of 5000i,500n,10000p,200000b,80000s

Output written on sympy-1.13.dev.xdv (2246 pages, 47913252 bytes).
Error: Process completed with exit code 1.
```

I am not sure what the problem is here. I tried looking it up, but could not find anything usefull. I came across and old issue (#24098) that had to do with sphinx, but this seems like a different problem.

Does anyone know what the problem is here and how it can be fixed?

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -491,6 +528,108 @@ def apply_support(self, loc, type="fixed"):
+            If type = "rotation"
+                - The unknown rotation jump multiplied by the elastic modulus and second moment as a symbol.
+            If type = "sliding"
+                - The unknown deflection jump multiplied by the elastic modulus and second moment as a symbol.
```
I think the latex if failing on this formatting.

moorepants replied:
Maybe just make it normal sentences in a paragraph.

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -491,6 +528,108 @@ def apply_support(self, loc, type="fixed"):
         else:
             return reaction_load, reaction_moment
 
+    def apply_hinge(self, loc, type="rotation"):
```
How many hinge types do you imagine there being? If only these two, why not make two separate methods. Then you wouldn't return different things depending on the kwarg.

e.g. `apply_rotation_hinge()` and `apply_sliding_hinge()`

Tom-van-Woudenberg replied:
I agree with the main reason that sliding hinges are not very common

mvg2002 replied:
These are the only two types I imagine being there. I will make two separate methods and then push again.

Tom-van-Woudenberg commented:
> I already added EI to the Symbol name, so when printing `b.rotation_jumps` instead of {P_7: 120} it would give {EI_P_7: 120}. I agree that for clarity, it is better to also add the EI to the variable name itself. I just pushed this change.

I was rethinking this. This double symbol is a bit strange. Couldn't you use your actual value of `EI` in the calculation and return the rotation jump itself? I think that's most useful for users.

Commit:
Split apply_hinge into apply_rotation_hinge and apply_sliding_hinge.

mvg2002 commented:
> @mvg2002, I was rethinking this one. The double symbol is a bit strange. Couldn't you use your actual value of `EI` in the calculation and return the rotation jump itself? I think that's most useful for users.

@Tom-van-Woudenberg, I will take a look at this.

Commit:
Fix code quality.

Commit:
Return actual rotation and deflection jump.

mvg2002 commented:
@Tom-van-Woudenberg, the properties `rotation_jumps` and `deflection_jumps` now return the actual values of the jumps. The method `_calculate_jumps` is used to divide the values following from the solution by EI. 

The `_calculate_jumps` method looks up the EI at the place of the hinge and divides by the correct value. This also works for beams with different I, made with the `join` method. 

While making this change I saw that it is not possible right now to create a beam with different E. I made an issue about this, visible here: #26628.

Tom-van-Woudenberg commented:
> @Tom-van-Woudenberg, the properties `rotation_jumps` and `deflection_jumps` now return the actual values of the jumps. The method `_calculate_jumps` is used to divide the values following from the solution by EI.
> 
> The `_calculate_jumps` method looks up the EI at the place of the hinge and divides by the correct value. This also works for beams with different I, made with the `join` method.
> 
> While making this change I saw that it is not possible right now to create a beam with different E. I made an issue about this, visible here: #26628.

@mvg2002 
Nice! However...
is it needed to calculate the combined term `self._EI_rotation_jumps`? Can you skip that step and directly calculate the jump itself?

mvg2002 commented:
> @mvg2002 Nice! However... is it needed to calculate the combined term `self._EI_rotation_jumps`? Can you skip that step and directly calculate the jump itself?

@Tom-van-Woudenberg, The `self._EI_rotation_jumps` and `self._EI_deflection_jumps` still need to be calculated, because these values are substituted into the load equation. Which is done in the last two lines here.
```
self._EI_rotation_jumps = dict(zip(rotation_jumps, rotation_solution))
self._EI_deflection_jumps = dict(zip(deflection_jumps, deflection_solution))
self._rotation_jumps = self._calculate_jumps(self._EI_rotation_jumps)
self._deflection_jumps = self._calculate_jumps(self._EI_deflection_jumps)
self._load = self._load.subs(self._reaction_loads)
self._load = self._load.subs(self._EI_rotation_jumps)
self._load = self._load.subs(self._EI_deflection_jumps)
```

This can be changed to: 
```
self._rotation_jumps = self._calculate_jumps(dict(zip(rotation_jumps, rotation_solution)))
self._deflection_jumps = self._calculate_jumps(dict(zip(deflection_jumps, deflection_solution)))
self._load = self._load.subs(self._reaction_loads)
self._load = self._load.subs(dict(zip(rotation_jumps, rotation_solution)))
self._load = self._load.subs(dict(zip(deflection_jumps, deflection_solution)))
```

The first code block has my preference, as this it is a bit more clear what is happening in my opinion.

Tom-van-Woudenberg commented:
> @Tom-van-Woudenberg, The `self._EI_rotation_jumps` and `self._EI_deflection_jumps` still need to be calculated, because these values are substituted into the load equation. Which is done in the last two lines here.
> 
> ```
> self._EI_rotation_jumps = dict(zip(rotation_jumps, rotation_solution))
> self._EI_deflection_jumps = dict(zip(deflection_jumps, deflection_solution))
> self._rotation_jumps = self._calculate_jumps(self._EI_rotation_jumps)
> self._deflection_jumps = self._calculate_jumps(self._EI_deflection_jumps)
> self._load = self._load.subs(self._reaction_loads)
> self._load = self._load.subs(self._EI_rotation_jumps)
> self._load = self._load.subs(self._EI_deflection_jumps)
> ```
> 
> This can be changed to:
> 
> ```
> self._rotation_jumps = self._calculate_jumps(dict(zip(rotation_jumps, rotation_solution)))
> self._deflection_jumps = self._calculate_jumps(dict(zip(deflection_jumps, deflection_solution)))
> self._load = self._load.subs(self._reaction_loads)
> self._load = self._load.subs(dict(zip(rotation_jumps, rotation_solution)))
> self._load = self._load.subs(dict(zip(deflection_jumps, deflection_solution)))
> ```
> 
> The first code block has my preference, as this it is a bit more clear what is happening in my opinion.

But what if the load equation itself has two seperate terms EI and rotation jump? So for example for some beam for which we add a hinge the load equation becomes $q = {A_v} \cdot {\left\langle x \right\rangle ^{ - 1}} + {A_m} \cdot {\left\langle x \right\rangle ^{ - 2}} + EI \cdot \varphi  \cdot {\left\langle {x - 2} \right\rangle ^{ - 3}} + {B_v}{\left\langle {x - 4} \right\rangle ^{ - 3}}$ instead of (as proposed by Justus) $q = {A_v} \cdot {\left\langle x \right\rangle ^{ - 1}} + {A_m} \cdot {\left\langle x \right\rangle ^{ - 2}} + \varphi  \cdot {\left\langle {x - 2} \right\rangle ^{ - 3}} + {B_v}{\left\langle {x - 4} \right\rangle ^{ - 3}}$. Then we only calculate and substitute $\phi$

mvg2002 commented:
> @mvg2002 But what if the load equation itself has two seperate terms EI and rotation jump? So for example for some beam for which we add a hinge the load equation becomes q=Av⋅⟨x⟩−1+Am⋅⟨x⟩−2+EI⋅φ⋅⟨x−2⟩−3+Bv⟨x−4⟩−3 instead of (as proposed by Justus) q=Av⋅⟨x⟩−1+Am⋅⟨x⟩−2+φ⋅⟨x−2⟩−3+Bv⟨x−4⟩−3. Then we only calculate and substitute φ

@Tom-van-Woudenberg I think I misunderstood your question earlier. You're proposal to directly calculate the jump itself should also be possible. I think in your way it is better understandable what is calculated. I will try to implement this.

Commit:
Replace calculation of EI times jump by immediate calculation of jump.

mvg2002 commented:
@Tom-van-Woudenberg I implemented the new calculations. The helper function `_get_I` finds the value of the second moment at the position of the hinge. In the line `self.apply_load(E * I * deflection_jump, loc, -4)`, the value of the deflection jump is multiplied by EI and added to the load equation. This is implemented in the same way for ratation jumps. The code is much cleaner this way, thank you for the valuable feedback.

Commit:
Raise NotImplementedError when joining beams with different E.

mvg2002 commented:
> Well done! I don't know whether Macaulay's method has been applied for beams with a varying E or I. So unless it was already there you might want to use a not implemented error here

@Tom-van-Woudenberg It was possible to join beams with different Elastic Modulus, but the code could not handle this and would generate wrong results. I added a NotImplementedError for this.

Tom-van-Woudenberg commented:
> @Tom-van-Woudenberg and @moorepants, the last feedback is implemented. Are there any other changes or improvements needed, or is it ready for approval?

@mvg2002 , I tried using it myself, see https://github.com/Tom-van-Woudenberg/Macaulay-various/blob/main/test_from_Justus_3.ipynb and https://github.com/Tom-van-Woudenberg/Macaulay-various/blob/main/test_from_apply_sliding_hinge.ipynb
I found one bug in `point_cflexure()`. That one doesn't seem to work with the hinges (although this specific example doesn't have a flexure point). I does work when there are no hinges. I don't know whether it was functional with the old `join_beam` thingy. If it's an existing issue, this can be merged.

moorepants commented:
If Tom has some examples in those notebooks, maybe they can be added as tests or doctests.

mvg2002 commented:
Thank you @Tom-van-Woudenberg for pointing this out. I did not come across this and it didn't raise any errors in the (doc)tests. I agree with @moorepants that this would be a good addition to the tests file. I will add this. 


The code of `point_cflexure`:
```
        moment_curve = Piecewise((float("nan"), self.variable<=0),
                (self.bending_moment(), self.variable<self.length),
                (float("nan"), True))

        points = solve(moment_curve.rewrite(Piecewise), self.variable,
                        domain=S.Reals)
        return points
```
Regarding the error. I expect that this errors occurs because there are still singularity functions with order -1 in the moment_curve as result of the hinge placements. I will take a look at this moment_curve equation and try to handle this error. 

I will also take a look at the other functions and determine if there are more errors like this possible.

Commit:
Fix point_cflexure error.

mvg2002 commented:
@Tom-van-Woudenberg I just pushed some changes that fix the error you got. I also added your example to the test file. 

 The problem in the code was twofold. 

The first part of the problem had to do with the singularity function of order -1. This gave a bending moment of infinity at the location of the hinge. This problem is solved by the `new_bending_moment` line. This line removes all singularity function with an order lower than 0 from the equation used in this method. 

The second part of the problem was already present in the old code. If a part of the beam had a constant bending moment of 0 the code gave an NotImplementedError. In your example the bending moment was 0 from x=0 to x=5. To handle this problem I added an try - except statement that gives a more relevant error message. 

The new code is as follows:

```
        new_bending_moment = sum(arg for arg in self.bending_moment().args if not arg.args[1].args[2] < 0)

        # To restrict the range within length of the Beam
        moment_curve = Piecewise((float("nan"), self.variable<=0),
                (new_bending_moment, self.variable<self.length),
                (float("nan"), True))
        try:
            points = solve(moment_curve.rewrite(Piecewise), self.variable,
                           domain=S.Reals)
        except NotImplementedError as e:
            if "An expression is already zero when" in str(e):
                raise NotImplementedError("This method cannot be used when a whole region of "
                                          "the bending moment line is equal to 0.")
            else:
                raise

        return points
```

Tom-van-Woudenberg commented:
> @Tom-van-Woudenberg I just pushed some changes that fix the error you got. I also added your example to the test file. 
> 
>  The problem in the code was twofold. 
> 
> The first part of the problem had to do with the singularity function of order -1. This gave a bending moment of infinity at the location of the hinge. This problem is solved by the `new_bending_moment` line. This line removes all singularity function with an order lower than 0 from the equation used in this method. 
> 
> The second part of the problem was already present in the old code. If a part of the beam had a constant bending moment of 0 the code gave an NotImplementedError. In your example the bending moment was 0 from x=0 to x=5. To handle this problem I added an try - except statement that gives a more relevant error message. 
> 
> The new code is as follows:
> 
> ```
>         new_bending_moment = sum(arg for arg in self.bending_moment().args if not arg.args[1].args[2] < 0)
> 
>         # To restrict the range within length of the Beam
>         moment_curve = Piecewise((float("nan"), self.variable<=0),
>                 (new_bending_moment, self.variable<self.length),
>                 (float("nan"), True))
>         try:
>             points = solve(moment_curve.rewrite(Piecewise), self.variable,
>                            domain=S.Reals)
>         except NotImplementedError as e:
>             if "An expression is already zero when" in str(e):
>                 raise NotImplementedError("This method cannot be used when a whole region of "
>                                           "the bending moment line is equal to 0.")
>             else:
>                 raise
> 
>         return points
> ```

@mvg2002 , i think you could think of a better name than `new_bending_moment`.
Furthermore, @moorepants suggested you added the examples to the docstring as well right? Could you do that?

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -210,6 +213,16 @@ def reaction_loads(self):
 
+    @property
+    def rotation_jumps(self):
+        """ Returns the rotation jumps in rotation hinges in a dictionary."""
```
It isn't so clear what this returns. Is a "jump" a symbol? expression? number? Making the docstring a bit more informative would help.

mvg2002 replied:
Changed the docstring to: 
"""
Returns the value for the rotation jumps in rotation hinges in a dictionary.
The rotation jump is the rotation (in degrees) in a rotation hinge. This can
be seen as a jump in the slope plot.
"""

Same for the deflection jump. Is it okay like this?

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -332,13 +345,29 @@ def boundary_conditions(self):
         >>> b.bc_slope = [(0, 1)]
         >>> b.boundary_conditions
-        {'deflection': [(0, 2)], 'slope': [(0, 1)]}
+        {'bending moment': [], 'deflection': [(0, 2)], 'shear force': [], 'slope': [(0, 1)]}
```
```suggestion
        {'bending_moment': [], 'deflection': [(0, 2)], 'shear_force': [], 'slope': [(0, 1)]}
```
I think making these keys valid Python identifiers may be best, for unknown future use (and tab completion in IDEs).

moorepants replied:
also matches the property names

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -470,6 +502,12 @@ def apply_support(self, loc, type="fixed"):
         """
         loc = sympify(loc)
+
+        if loc in self._applied_rotation_hinges and type == "fixed":
```
If loc is a floating point value, it is possible `loc in [list of floats]` will not find it due to ineact matches on floats.

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -491,6 +529,171 @@ def apply_support(self, loc, type="fixed"):
+
+    def apply_rotation_hinge(self, loc):
+        """
+        This method applies a rotation hinge in a particular beam object.
```
```suggestion
        This method applies a rotation hinge at a single location on the beam.
```

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -491,6 +529,171 @@ def apply_support(self, loc, type="fixed"):
+        >>> b.reaction_loads
+        {M_15: -75/2, R_0: 0, R_10: 40, R_15: -5}
+        >>> b.rotation_jumps
+        {P_12: -1875/(16*E*I), P_5: 9625/(24*E*I)}
```
What are the units of a jump? "loads" are forces and moments. Is a jump a distance?

mvg2002 replied:
A rotation jump is a rotation in degrees and a deflection jump is a distance in meters. This is now also added to the docstring.

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -491,6 +529,171 @@ def apply_support(self, loc, type="fixed"):
+        #Check for unusual hinge placement
+        warning_message = 'It is highly unusual to place a rotation hinge and a sliding hinge at the same location.'
+        if loc in self._applied_sliding_hinges:
+            warnings.warn(warning_message)
```
If it is valid to do so, I don't think we need this warning. If it isn't mathematically valid, then throw an error instead.

mvg2002 replied:
It is mathematically valid to do this. I will remove the warning.

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -491,6 +529,171 @@ def apply_support(self, loc, type="fixed"):
+            warnings.warn(warning_message)
+
+        #Check for other errors
+        if loc == 0:
```
In SymPy checking if zero is `if loc.is_zero`.

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -491,6 +529,171 @@ def apply_support(self, loc, type="fixed"):
+        #Check for other errors
+        if loc == 0:
+            raise ValueError('Cannot place hinge at the beginning of the beam.')
+        if loc == self.length:
```
I don't think this works for arbitrary expressions. If `loc` and `self.length` are symbolic expressions, then checking if they are equal is nontrivial. `sm.simplify(loc - self.length) == 0` is a possibly but not ideal to use simplify() in library code. It is ok to use the "garbage in -> garbage out" philosophy and not check for all possible bad inputs.

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -491,6 +529,171 @@ def apply_support(self, loc, type="fixed"):
+        if any(loc == arg[1] and arg[2] == -2 for arg in self._applied_loads):
+            raise ValueError('Cannot place rotation hinge at the location of a moment load.')
+        if any(loc == support[0] and support[1] == 'fixed' for support in self._applied_supports):
+            raise ValueError('Cannot place rotation hinge at the location of a fixed support. Change fixed support to pin.')
```
All of these checks will break down if the variables are symbolic expressions.

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -491,6 +529,171 @@ def apply_support(self, loc, type="fixed"):
+        if any(loc == support[0] and support[1] == 'fixed' for support in self._applied_supports):
+            raise ValueError('Cannot place rotation hinge at the location of a fixed support. Change fixed support to pin.')
+
+        rotation_jump = Symbol('P_'+str(loc))
```
This works nicely if `loc` is an integer, but what about if it is a float or a symbolic expression?

mvg2002 replied:
This should not be a problem. This is symbol is returned and the user can give it any other name they want. For example in this code:
```
b = Beam(13, 20, 20)
r0, m0 = b.apply_support(0, type="fixed")
w333 = b.apply_sliding_hinge(10/3)
r13 = b.apply_support(13, type="pin")
b.apply_load(-10, 5, -1)
b.solve_for_reaction_loads(r0, m0, r13)
print(b.deflection_jumps)
print(b.deflection_jumps[w333])
```
The output is: 
`{W_3.33333333333333: -443/30}`
`-443/30`

The string of the location is there so the user can know which hinge is meant when there are multiple hinges in the beam. The value op the jump can be called by using the name they gave to the symbol of the jump. This works the same way when a symbolic expression is used to describe the location of the hinge. 

```
b = Beam(L, 20, 20)
r0, m0 = b.apply_support(0, type="fixed")
wL3 = b.apply_sliding_hinge(L/3)
rL = b.apply_support(L, type="pin")
b.apply_load(-10, L/2, -1)
b.solve_for_reaction_loads(r0, m0, rL)
print(b.deflection_jumps)
print(b.deflection_jumps[wL3])
```
Output: 
`{W_L/3: -11*L**3/1920}`
`-11*L**3/1920`

moorepants replied:
Yes, symbols can be named like this, but such symbol names can break code printers and may have other undesired effects when used in other parts of sympy. Most other functions that need to create symbol names for the user keep them simple (like `C1,C2,...` in `dsolve()`). I see that we are already doing this kind of string creation:

```
grep -r Symbol sympy/physics/continuum_mechanics/beam.py 
from sympy.core import S, Symbol, diff, symbols
    def __init__(self, length, elastic_modulus, second_moment, area=Symbol('A'), variable=Symbol('x'), base_char='C'):
            A Symbol or value representing the Beam's length.
        area : Symbol/float
        variable : Symbol, optional
            A Symbol object that will be used as the variable along the beam
            curve. By default, it is set to ``Symbol('x')``.
        to ``Symbol('x')``, but this property is mutable.
        if isinstance(v, Symbol):
            raise TypeError("""The variable should be a Symbol object.""")
        Symbol or tuple of Symbol
            - Symbol(reaction_force) if type = "pin" or "roller"
            - Symbol(reaction_force), Symbol(reaction_moment) if type = "fixed"
            reaction_load = Symbol('R_'+str(loc))
            reaction_load = Symbol('R_'+str(loc))
            reaction_moment = Symbol('M_'+str(loc))
        h = Symbol('h')     # Force due to hinge
        C1 = Symbol('C1')
        C2 = Symbol('C2')
        C3 = Symbol('C3')
        C4 = Symbol('C4')
        C3 = Symbol('C3')
        C4 = Symbol('C4')
        C3 = Symbol('C3')
        C4 = Symbol('C4')
            Python dictionary containing Symbols as key and their
        for sym in shear_stress.atoms(Symbol):
            Python dictionary containing Symbols as key and their
        for sym in shear_force.atoms(Symbol):
            Python dictionary containing Symbols as key and their
        for sym in bending_moment.atoms(Symbol):
            Python dictionary containing Symbols as key and their
        for sym in slope.atoms(Symbol):
            Python dictionary containing Symbols as key and their
        for sym in deflection.atoms(Symbol):
               Python dictionary containing Symbols as key and their
        for sym in self.deflection().atoms(Symbol):
        C3 = Symbol('C3')
        C4 = Symbol('C4')
               Python dictionary containing Symbols as key and their
            for sym in self._ild_reactions[reaction].atoms(Symbol):
        for sym in self._length.atoms(Symbol):
               Python dictionary containing Symbols as key and their
        for sym in self._ild_shear.atoms(Symbol):
        for sym in self._length.atoms(Symbol):
               Python dictionary containing Symbols as key and their
        for sym in self._ild_moment.atoms(Symbol):
        for sym in self._length.atoms(Symbol):
        # checking whether length is an expression in terms of any Symbol.
            l = list(self.length.atoms(Symbol))
            # assigning every Symbol a default value of 10
                 area, variable=Symbol('x')):
            A Symbol or value representing the Beam's length.
        variable : Symbol, optional
            A Symbol object that will be used as the variable along the beam
            curve. By default, it is set to ``Symbol('x')``.
            reaction_load = Symbol('R_'+str(loc))
            reaction_load = Symbol('R_'+str(loc))
            reaction_moment = Symbol('M_'+str(loc))
        C1 = Symbol('C1')
        C2 = Symbol('C2')
        C_i = Symbol('C_i')
        for sym in shear_force[dir_num].atoms(Symbol):
            Python dictionary containing Symbols as key and their
        for sym in bending_moment[dir_num].atoms(Symbol):
            Python dictionary containing Symbols as key and their
        for sym in slope[dir_num].atoms(Symbol):
            Python dictionary containing Symbols as keys and their
        for sym in deflection[dir_num].atoms(Symbol):
            Python dictionary containing Symbols as keys and their
               Python dictionary containing Symbols as key and their
        for sym in shear_stress[dir_num].atoms(Symbol):
            Python dictionary containing Symbols as key and their
```

So, it is fine to continue to do so.

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -491,6 +529,171 @@ def apply_support(self, loc, type="fixed"):
+        if any(loc == arg[1] and arg[2] == -1 for arg in self._applied_loads):
+            raise ValueError('Cannot place sliding hinge at the location of a point load.')
+        if any(loc == support[0] for support in self._applied_supports):
+            raise ValueError('Cannot place sliding hinge at the location of a support.')
```
given that all the code above this line is duplicate in your two hinge methods, you can make a helper method that does these checks and then both hinge methods call it, deduplicating code.

Commit:
Improve docstring rotation_jumps and defelction_jumps.

Commit:
Change bending moment and shear force.

Commit:
Change new bending moment to real bending moment.

Commit:
Change docstring apply hinges.

mvg2002 commented:
> @mvg2002 , i think you could think of a better name than `new_bending_moment`. And could you add an issue for a moment line region of 0?

@Tom-van-Woudenberg, I changed the name to `real_bending_moment`  and made the following issue: #26634

Commit:
Remove warnings.

mvg2002 commented:
@moorepants Regarding the failing checks for symbolic and float inputs. I had never heard about the "garbage in -> garbage out" philosophy, but I understand the idea of it. 

Using this idea I think the following checks can be removed:
```
        if loc in self._applied_rotation_hinges:
            raise ValueError('Cannot place two rotation hinges at the same location.')

        if loc == 0:
            raise ValueError('Cannot place hinge at the beginning of the beam.')

        if loc == self.length:
            raise ValueError('Cannot place hinge at the end of the beam.')
```

There are also some checks that test less trivial faulty placements: 
```
        if any(loc == arg[1] and arg[2] == -2 for arg in self._applied_loads):
            raise ValueError('Cannot place rotation hinge at the location of a moment load.')

        if any(loc == support[0] and support[1] == 'fixed' for support in self._applied_supports):
            raise ValueError('Cannot place rotation hinge at the location of a fixed support. Change fixed support to pin.')
```

These lines check for faults a user kan make by mistake. Do you think these tests can also be removed or is it better to change these checks in a certain way that they can handle float and symbolic inputs?

If you think it is better to keep the tests, do you maybe have some hints on how to do this? I tried a few possible solutions but they did not work.

Commit:
Change variable name to non singular bending moment.

mvg2002 commented:
> > @Tom-van-Woudenberg, I changed the name to `real_bending_moment` and made the following issue: #26634
> 
> Still a bit vague? What about `non_zero_moment`?

It is not accually the non zero moment. The moment can still be zero at points or even regions in the bending moment plot. The difference is that the bending moment equation no more containes singularity functions of orders lower than 0. 

I changed it to `non_singular_bending_moment` as it is the bending moment equation without the singularity functions that become infinity at one single point. I also added a comment for more clarity. 

```
#Removes the singularity functions of order < 0 from the bending moment equation used in this method
non_singular_bending_moment = sum(arg for arg in self.bending_moment().args if not arg.args[1].args[2] < 0)
```

Commit:
Remove placement checks.

mvg2002 commented:
@Tom-van-Woudenberg and @moorepants, I just pushed a new version without the placement checks when applying a hinge. This is done because of the 'garbage in -> garbage out' phylosophy.

It would be nice if there was a possibilty to place a hinge "just" before or "just" after a support or load. This was not possible in the previous implementation and overshoots the goal of this pull request. I made an issue for this possible improvement, visible here: #26639. 

The rest of the given feedback is implemented, are there any more changes necessary?

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -210,6 +213,24 @@ def reaction_loads(self):
+    def rotation_jumps(self):
+        """
+        Returns the value for the rotation jumps in rotation hinges in a dictionary.
+        The rotation jump is the rotation (in degrees) in a rotation hinge. This can
```
My guess is that this would be radians, not degrees. The angle computed from a slope with atan() gives radians.

mvg2002 replied:
You are right, I changed it

moorepants commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -370,7 +407,7 @@ def join(self, beam, via="fixed"):
             States the way two Beam object would get connected
             - For axially fixed Beams, via="fixed"
-            - For Beams connected via hinge, via="hinge"
+            - For Beams connected via rotaion hinge, via="hinge"
```
```suggestion
            - For Beams connected via rotation hinge, via="hinge"
```

moorepants commented:
I left a spelling suggestion and are you sure it is degrees not radians?

Tom-van-Woudenberg commented on .mailmap:
```diff
@@ -1599,6 +1599,7 @@ mao8 <thisisma08@gmail.com>
 mohajain <mohajain99@gmail.com> mohajain <45903778+mohajain@users.noreply.github.com>
 mohammedouahman <simofun85@gmail.com>
 mohit <39158356+mohitacecode@users.noreply.github.com> mohit <42018918+mohitshah3111999@users.noreply.github.com>
+mvg2002 <168417631+mvg2002@users.noreply.github.com>
```
shouldn't you add your own mail address here? @mvg2002

oscarbenjamin replied:
It should be like:
```
Real Name <myemail@example.com> mvg2002 <168417631+mvg2002@users.noreply.github.com>
```
Then the AUTHORS file will end up showing `Real Name <myemail@example.com>`.

mvg2002 replied:
My real name and email were already in the mailmap from an old pull request. I had to add this git email as well to pass the check in this PR. I changed it now to:
`Real Name <myemail@example.com> mvg2002 <168417631+mvg2002@users.noreply.github.com>`

Tom-van-Woudenberg replied:
@mvg2002 it's better to use your personal email. You'll loose access to your student mail at some point.

mvg2002 replied:
@Tom-van-Woudenberg I had not thought about that, I changed it.

Commit:
Unit mistake, spelling mistake and mailmap fix.

Commit:
AUTORS vs mailmap problem fix.

mvg2002 committed:
Merge branch 'master' into Dev_hinge_beams

Commit:
mailmap fix.

Commit:
Merge remote-tracking branch 'origin/Dev_hinge_beams' into Dev_hinge_beams

Commit:
mailmap fix.

moorepants commented:
This is going in! Nice work :)

mvg2002 commented:
Perfect, thank you!

Tom-van-Woudenberg commented:
@mvg2002 , https://docs.sympy.org/latest/modules/physics/continuum_mechanics/beam_problems.html this page now contains some outdated examples

mvg2002 commented:
> @mvg2002 , https://docs.sympy.org/latest/modules/physics/continuum_mechanics/beam_problems.html this page now contains some outdated examples

@Tom-van-Woudenberg, These examples are indeed outdated. I will take a look at this and try to update the code.

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -1115,14 +1168,23 @@ def point_cflexure(self):
         [10/3]
         """
+        #Removes the singularity functions of order < 0 from the bending moment equation used in this method
+        non_singular_bending_moment = sum(arg for arg in self.bending_moment().args if not arg.args[1].args[2] < 0)
```
@mvg2002 , wouldn't it make sense if the singularity function would never return $\infty$?

mvg2002 replied:
This has to do with the definition of negative order singularity functions, take for example $< x - a > ^ {-1}$. This function can be in the bending moment equation as result of a rotation hinge at $x = a$. 

The method `point_cflexure` searches for points where the bending moment is equal to 0. As there is a hinge located at  $x = a$ the bending moment should be 0 at that point and the method should return $x = a$ as a point where the bending moment is 0. 

But when $< x - a > ^ {-1}$ is still present in the bending moment equation, evaluating it at $x = a$ would return $\infty$. The method would not give $x = a$ as a point where the bending moment is 0. That is why for this calculation singularity functions of negative order are removed, they don't do anything except set the bending moment to infinity in the hinges. 

The singularity function should return $\infty$ when $x = a$, as follows from their mathematical definition. This should not be changed in SymPy. There might be other people using the singularity functions and they expect them to return $\infty$ when $x = a$

Tom-van-Woudenberg replied:
I agree that we shouldn't change it everywhere, but wouldn't the mechanics application benefit from a definition without $\infty$?

mvg2002 replied:
@Tom-van-Woudenberg, For mechanical applications of the singularity function returning $\infty$ is never needed and can only complicate things. I agree that the mechanics application would benifit from a definition without $\infty$.

Tom-van-Woudenberg replied:
@moorepants, would there be a place in SymPy for mechanics-specific singularity functions?
